listenbrainz/mbid_mapping_writer/job_queue.py

```diff
@@ -86,11 +108,60 @@ def load_legacy_listens(self):
             target=_add_legacy_listens_to_queue, args=(self,))
         self.legacy_load_thread.start()
 
+    def fetch_and_queue_listens(self, query, args):
+        """ Fetch and queue legacy and recheck listens """
+
+        msb_query = """SELECT gid AS recording_msid
+                            , rj.data->>'title' AS track_name
+                            , rj.data->>'artist' AS artist_name
+                         FROM recording r
+                         JOIN recording_json rj
+                           ON r.data = rj.id
+                       WHERE gid in :msids"""
+
+        count = 0
+        msids = []
+        with timescale.engine.connect() as connection:
+            curs = connection.execute(sqlalchemy.text(query), args)
+            for row in curs.fetchall():
+                msids.append(row["recording_msid"])
+
+        if len(msids) == 0:
+            return 0
+
+        with msb_db.engine.connect() as connection:
+            curs = connection.execute(sqlalchemy.text(msb_query), msids=tuple(msids))
+            while True:
+                result = curs.fetchone()
+                if not result:
+                    break
+
+                self.queue.put(JobItem(LEGACY_LISTEN, [{"data": {"artist_name": result[2],
+                                                                 "track_name": result[1]},
+                                                        "recording_msid": result[0],
+                                                        "legacy": True}]))
+                count += 1
+
+        return count
+
     def add_legacy_listens_to_queue(self):
         """Fetch more legacy listens from the listens table by doing an left join
            on the matched listens, finding the next chunk of legacy listens to look up.
            Listens are added to the queue with a low priority."""
 
+        legacy_query = """SELECT data->'track_metadata'->'additional_info'->>'recording_msid'::TEXT AS recording_msid
```

Comment:
This query seemed weird at first look.  Maybe a comment to tell its to select not yet mapped msids? Also, would it make to use a WHERE EXISTS or WHERE IN?

Reply:
:+1: